tsl/src/compression/compression.c

```diff
@@ -1450,3 +1452,47 @@ compression_get_toast_storage(CompressionAlgorithms algorithm)
 		elog(ERROR, "invalid compression algorithm %d", algorithm);
 	return definitions[algorithm].compressed_data_storage;
 }
+
+/* Get relstats from compressed chunk and insert into relstats for the
+ * corresponding uncompressed parent.
+ */
+extern void
+update_compressed_chunk_relstats(Oid uncompressed_relid, Oid compressed_relid)
+{
+	double rowcnt;
+	int comp_pages, uncomp_pages, comp_visible, uncomp_visible;
+	float comp_tuples, uncomp_tuples, out_tuples;
+	Chunk *uncompressed_chunk = ts_chunk_get_by_relid(uncompressed_relid, true);
+	Chunk *compressed_chunk = ts_chunk_get_by_relid(compressed_relid, true);
+
+	if (uncompressed_chunk->table_id != uncompressed_relid ||
+		uncompressed_chunk->fd.compressed_chunk_id != compressed_chunk->fd.id ||
+		compressed_chunk->table_id != compressed_relid)
+	{
+		ereport(ERROR,
+				(errcode(ERRCODE_INTERNAL_ERROR),
+				 errmsg("mismatched chunks for relstats update %d %d",
+						uncompressed_relid,
+						compressed_relid)));
+	}
+
+	/* we need page info from compressed relid */
+	capture_pgclass_stats(compressed_relid, &comp_pages, &comp_visible, &comp_tuples);
+	rowcnt = (double) ts_get_before_compression_row_count(uncompressed_chunk->fd.id);
+	capture_pgclass_stats(uncompressed_relid, &uncomp_pages, &uncomp_visible, &uncomp_tuples);
+
+	/* Before compressing a chunk in 2.0, we save its stats. Prior
+	 * releases do not support this. So the stats on uncompressed relid
+	 * could be invalid. In this case, do the best that we can.
+	 * Alternately, the stats could match, don't update in that case.
+	 */
+	if (uncomp_tuples == 0)
```

Comment:
The comment and the condition do not match. It doesn't update if the number of uncompressed tuples is set. However, it doesn't check if the number of uncompressed tuples is correct or not. What do I miss?

Reply:
We have no way of verifying the number of uncompressed tuples without actually decompressing the chunk.

Reply:
Then I suggest to improve the comment. E.g.,
```suggestion
	 * Alternately, the stats could already exist, don't update in that case.
	 */
	if (uncomp_tuples == 0)
```

Reply:
modified.